modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc

```diff
@@ -0,0 +1,535 @@
+                WHERE f.PhysiologicalFileID=:PFID AND f.FileType="json"',
+            ['PFID' => $this->_physioFileID]
+        );
+
+        $annotationLabels = $db->pselect(
+            'SELECT * FROM physiological_annotation_label',
+            ['PFID' => $this->_physioFileID]
+        );
+
+        $this->_data = [
+            'instances' => $annotationInstance,
+            'metadata'  => $annotationMetadata,
+            'labels'    => $annotationLabels,
+        ];
+    }
+
+    /**
+     * Get data for the Electrophysiological file annotations
+     *
+     * @return array The data array
+     */
+    function getData(): array
+    {
+        return $this->_data;
+    }
+
+    /**
+     * Updates annotation tables when there is a POST request.
+     * Will add new derivative files if none exist for the given instance.
+     * Will either add new annotations or update existing ones.
+     *
+     * @param array $instance_data Instance data
+     * @param array $metadata      Metadata
+     * @param int   $instance_id   InstanceID
+     * @param int   $parameter_id  ParameterID
+     *
+     * @return void
+     */
+    function update(
+        array $instance_data,
+        array $metadata,
+        int $instance_id,
+        int $parameter_id
+    ): void {
+
+        $factory = \NDB_Factory::singleton();
+        $user    = $factory->user();
+        $db      = $factory->database();
+
+        if ($user->hasPermission('electrophysiology_browser_edit_annotations')) {
+
+            //If the label is new, add to annotation label table
+            //and get label ID
+            $labelID = $db->pselectOne(
+                "SELECT AnnotationLabelID
+                    FROM physiological_annotation_label
+                    WHERE LabelName=:label",
+                ['label' => $instance_data['label_name']]
+            );
+            if (empty($labelID)) {
+                $data = [
+                    'LabelName'        => $instance_data['label_name'],
+                    'LabelDescription' => $instance_data['label_description']
+                ];
+                $db->insert("physiological_annotation_label", $data);
+                $labelID = $db->pselectOne(
+                    "SELECT AnnotationLabelID
+                        FROM physiological_annotation_label
+                        WHERE LabelName=:label",
+                    ['label' => $instance_data['label_name']]
+                );
+            }
+
+            //If no derivative files exist, must create new files
+            $annotationFID = $db->pselect(
+                "SELECT AnnotationFileID
+                    FROM physiological_annotation_file
+                    WHERE PhysiologicalFileID=:PFID",
+                ['PFID' => $this->_physioFileID]
+            );
+
+            //Get data from POST request
+            $metadata = [
+                'Description' => $metadata['description'],
+                'Sources'     => $metadata['sources'],
```

Comment:
Shouldn't this be $electrophysioFile?

Reply:
No. In the BIDS specifications there are two fields: `Sources` and `IntendedFor`. IntendedFor is defined as: `REQUIRED. Link to the data file for which the annotations are intended.` and  Sources is defined as `RECOMMENDED. Relative link or description of the file(s) used to make the annotations.`. For the IntendedFor entry, I link this data to a physiological file ID in the database.

Reply:
Following our discussion, I think 'Sources' is not necessary and 'this can be removed.